pkg/controller/volume/attachdetach/util/util.go

```diff
@@ -249,3 +274,109 @@ func ProcessPodVolumes(pod *v1.Pod, addVolumes bool, desiredStateOfWorld cache.D
+		if err != nil {
+			return nil, err
+		}
+	}
+	return spec, nil
+}
+
+func isCSIMigrationSupportedOnNode(nodeName types.NodeName, spec *volume.Spec, vpm *volume.VolumePluginMgr, csiMigratedPluginManager volume.CSIMigratedPluginManager) (bool, error) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationInTreeOff) {
+		// All nodes are expected to have migrated CSI plugins installed and
+		// configured when CSIMigrationInTreeOff is enabled. CSI migration is
+		// supported even if there is version skew between managers and node.
+		return true, nil
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) {
+		// If CSIMigration is disabled, CSI migration paths will not be taken for
+		// the node. If CSINodeInfo is disabled, checking of installation status
+		// of a migrated CSI plugin cannot be performed. Therefore stick to
+		// in-tree plugins.
+		return false, nil
+	}
+
+	if len(nodeName) == 0 {
+		return false, errors.New("nodeName is empty")
+	}
+
+	kubeClient := vpm.Host.GetKubeClient()
+	if kubeClient == nil {
+		// Don't handle the controller/kubelet version skew check and fallback
+		// to just checking the feature gates. This can happen if
+		// we are in a standalone (headless) Kubelet
+		return true, nil
+	}
+
+	adcHost, ok := vpm.Host.(volume.AttachDetachVolumeHost)
+	if !ok {
+		// Don't handle the controller/kubelet version skew check and fallback
+		// to just checking the feature gates. This can happen if
+		// "enableControllerAttachDetach" is set to true on kubelet
+		return true, nil
+	}
+
+	if adcHost.CSINodeLister() == nil {
+		return false, errors.New("could not find CSINodeLister in attachDetachController")
+	}
+
+	csiNode, err := adcHost.CSINodeLister().Get(string(nodeName))
+	if err != nil {
+		return false, err
+	}
+
+	ann := csiNode.GetAnnotations()
+	if ann == nil {
+		return false, nil
+	}
+
+	mpa := ann[v1.MigratedPluginsAnnotationKey]
+	tok := strings.Split(mpa, ",")
+	mpaSet := sets.NewString(tok...)
+
+	pluginName, err := csiMigratedPluginManager.GetInTreePluginNameFromSpec(spec.PersistentVolume, spec.Volume)
+	if err != nil {
+		return false, err
+	}
+
+	if len(pluginName) == 0 {
+		// Could not find a plugin name from translation directory, assume not translated
+		return false, nil
+	}
+
+	isMigratedOnNode := mpaSet.Has(pluginName)
+
+	if isMigratedOnNode {
+		installed := false
+		driverName, err := csiMigratedPluginManager.GetCSINameFromInTreeName(pluginName)
+		if err != nil {
+			return isMigratedOnNode, err
+		}
+		for _, driver := range csiNode.Spec.Drivers {
+			if driver.Name == driverName {
+				installed = true
+				break
+			}
+		}
+		if !installed {
+			return true, fmt.Errorf("in-tree plugin %s is migrated on node %s but driver %s is not installed", pluginName, string(nodeName), driverName)
+		}
+	}
+
```

Comment:
not sure if this is github cod review glitching but is there supposed to be a final return in this func (I don't see one)?

Reply:
There is the final return at L381: https://github.com/kubernetes/kubernetes/pull/83098/files#diff-bed922cbf0ba8772ff72617e43367c89R381